worker/raft/rafttransport/worker.go

```diff
@@ -0,0 +1,336 @@
+}
+
+type dialResult struct {
+	conn net.Conn
+	err  error
+}
+
+// Kill is part of the worker.Worker interface.
+func (w *Worker) Kill() {
+	w.catacomb.Kill(nil)
+}
+
+// Wait is part of the worker.Worker interface.
+func (w *Worker) Wait() error {
+	return w.catacomb.Wait()
+}
+
+// dialRaw dials a new TLS connection to the controller identified
+// by the given address. The address is expected to be the stringified
+// tag of a controller machine agent. The resulting connection is
+// appropriate for use as Dialer.DialRaw.
+func (w *Worker) dialRaw(address raft.ServerAddress, timeout time.Duration) (net.Conn, error) {
+	tag, err := names.ParseMachineTag(string(address))
+	if err != nil {
+		return nil, net.InvalidAddrError(err.Error())
+	}
+
+	// Give precedence to the worker dying.
+	select {
+	case <-w.catacomb.Dying():
+		return nil, w.errDialWorkerStopped()
+	default:
+	}
+
+	ctx := context.Background()
+	if timeout != 0 {
+		var cancel context.CancelFunc
+		ctx, cancel = context.WithTimeout(ctx, timeout)
+		defer cancel()
+	}
+
+	resultCh := make(chan dialResult)
+	req := dialRequest{
+		ctx:    ctx,
+		tag:    tag,
+		result: resultCh,
+	}
+	select {
+	case w.dialRequests <- req:
+	case <-ctx.Done():
+		return nil, dialRequestTimeoutError{}
+	case <-w.catacomb.Dying():
+		return nil, w.errDialWorkerStopped()
+	}
+
+	select {
+	case res := <-resultCh:
+		return res.conn, res.err
+	case <-ctx.Done():
+		return nil, dialRequestTimeoutError{}
+	case <-w.catacomb.Dying():
+		return nil, w.errDialWorkerStopped()
+	}
+}
+
+func (w *Worker) errDialWorkerStopped() error {
+	err := w.catacomb.Err()
+	if err != nil && err != w.catacomb.ErrDying() {
+		return dialWorkerStoppedError{err}
+	}
+	return dialWorkerStoppedError{
+		errors.New("worker stopped"),
+	}
+}
+
+func (w *Worker) loop() error {
+	h := NewHandler(w.connections, w.catacomb.Dying())
+	w.config.Mux.AddHandler("GET", w.config.Path, &ControllerHandler{
+		Mux:     w.config.Mux,
+		Handler: h,
+	})
+	defer w.config.Mux.RemoveHandler("GET", w.config.Path)
+
+	for {
+		select {
+		case <-w.catacomb.Dying():
+			return w.catacomb.ErrDying()
+		case req := <-w.dialRequests:
+			go w.handleDial(req)
+		}
+	}
+}
+
+func (w *Worker) apiserverDetailsChanged(topic string, details pubsubapiserver.Details, err error) {
+	if err != nil {
+		// This should never happen, so treat it as fatal.
+		w.catacomb.Kill(errors.Annotate(err, "apiserver details callback failed"))
+		return
+	}
+	w.mu.Lock()
+	defer w.mu.Unlock()
+	w.servers = details
+	close(w.serversUpdated)
+	w.serversUpdated = make(chan struct{})
+}
+
+func (w *Worker) handleDial(req dialRequest) {
+	addrs := w.serverAddresses(req.ctx, req.tag)
+	if addrs == nil {
+		// context expired or canceled
+		return
+	}
+
+	dial := func() (net.Conn, error) {
+		// Dial an api.Connection first, so we get the
+		// right address to connect to. The multi-address
+		// connection logic should be refactored so we
+		// can do without this.
```

Comment:
does this still need to exist if we are coordinating the addresses in raft config already?